leaf-ui/rappid-extensions/BackboneModelsIntention.js

```diff
@@ -7,61 +8,404 @@
+            if (constraint.get('type') == 'A' && constraint.get('srcID') === this.cid) {
+                graph.get('constraints').splice(i, 1); 
+            } else {
+                i++;
+            }
+        }
+    }, 
+ 
+    /**
+     * Adds a new FunctionSegmentBBM to the end of this Intention's EvolvingFunction's
+     * function list.
+ 
+     * Also adds a corresponding Constraint object reprenting an absolute constraint
+     * into the global graph variable.
+     *
+     *This function should only be used to add new function
+     * segments for user defined functions
+     *
+     * @param {String} funcType
+     *   ex: 'C'
+     * @param {String} refEvidencePair
+     *   ex: '0000'
+     * @param {Integer} startTime
+     */
+    addUserDefinedSeg: function(funcType, refEvidencePair, startTime){
+
+        var len = this.getFuncSegments().length;
+        var startCheck = this.getFuncSegments()[len - 1].get('startTP'); // Get last value in list 
+        if (startCheck == '0') { // If previous segment is at 0 then next one is at A
+            var start = 'A';
+        } else { // Otherwise, for following segment increase number by one 
+            var start = String.fromCharCode(startCheck.charCodeAt(0) + 1); 
+        }
+ 
+        // Create new FunctionSegmentBBM and adds it to functionSegList 
+        var new_model = new FunctionSegmentBBM({type: funcType, refEvidencePair: refEvidencePair, startTP: start, startAT: myNull}); 
+        this.getFuncSegments().push(new_model); 
+        //graph.get('constraints').push(new ConstraintBBM({type: 'A', srcID: this.get('cid'), srcRefTP: start, destID: null, destRefTP: null}));
+ 
+    }, 
+    /**
+     * Sets the refEvidencePair for the FunctionSegmentBBMs in the
+     * EvolvingFunction for the Intention
+     *
+     * This function will only be called for I, D, RC, MP, MN functions
+     * @param {String} satValue
+     * ex: '0000'
+     */
+    // TODO: i think we are going to have to change this function when we add views for the FunctionSegmentBBMs
+    setMarkedValueToFunction: function(satValue) {
+
+        if (this.get('evolvingFunction') != null) {
+            var funcType = this.get('evolvingFunction').get('type'); 
+        } else { 
+            var funcType = null; 
+        }
+ 
+        var len = this.getFuncSegments().length;
+        this.getFuncSegments()[len - 1].set('refEvidencePair', satValue); // Set refEvidencePair for last value 
+ 
+        if (funcType == 'MP' || funcType == 'MN') {
+            this.getFuncSegments()[0].set('refEvidencePair', satValue); // Set refEvidencePair for initial value 
+        }
+    }, 
+ 
+    /**
+     * Sets the function type and marked value for the
+     * last FunctionSegmentBBM for the Intention's EvolvingFunction
+     *
+     * @param {String} funcValue
+     */
+    setUserDefinedSegment: function(funcValue) {
+
+        if (this.getFuncSegments() != null) {
```

Comment:
logic:
var funcSegments = this.getFuncSegments()
if (funcSegments != null){
get length and do all your stuff in here
} no else needed